src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1371,24 +1371,30 @@ public override BoundNode VisitLocalDeclaration(BoundLocalDeclaration node)
 
             TypeWithAnnotations type = local.TypeWithAnnotations;
             TypeWithState valueType;
+            bool inferredType = node.InferredType;
             if (local.IsRef)
             {
                 valueType = VisitRefExpression(initializer, type);
             }
             else
             {
-                bool inferredType = node.InferredType;
-                valueType = VisitOptionalImplicitConversion(initializer, targetTypeOpt: inferredType ? default : type, useLegacyWarnings: true, trackMembers: true, AssignmentKind.Assignment);
-                if (inferredType)
+                valueType = VisitOptionalImplicitConversion(initializer, targetTypeOpt: inferredType ? default : type, useLegacyWarnings: true, trackMembers: true, AssignmentKind.Assignment);         
+            }
+
+            if (inferredType)
+            {
+                if (valueType.HasNullType)
                 {
-                    if (valueType.HasNullType)
-                    {
-                        Debug.Assert(type.Type.IsErrorType());
-                        valueType = type.ToTypeWithState();
-                    }
+                    Debug.Assert(type.Type.IsErrorType());
+                    valueType = type.ToTypeWithState();
+                }
 
-                    type = valueType.ToTypeWithAnnotations();
-                    _variableTypes[local] = type;
+                type = valueType.ToTypeWithAnnotations();
+                _variableTypes[local] = type;
+
+                if (node.DeclaredTypeOpt != null)
+                {
+                    SetAnalyzedNullability(node.DeclaredTypeOpt, new VisitResult(valueType, type), true);
```

Comment:
>SetAnalyzedNullability(node.DeclaredTypeOpt, new VisitResult(valueType, type), true) [](start = 20, length = 84)

Is this necessary? (What is the observed effect?)

Reply:
In the semantic model, you can query the `var` of the declaration and get the type. This ensures we correctly report the nullability in that case.